torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -199,17 +210,17 @@ void TensorPipeAgent::pipeWrite(
     const std::shared_ptr<tensorpipe::Pipe>& pipe,
     Message&& rpcMessage,
     std::function<void(const tensorpipe::Error&)> fn) {
-  TensorPipeEntry tpEntry = tensorpipeSerialize(rpcMessage);
-  tensorpipe::Message tpMessage = std::move(tpEntry.message);
+  tensorpipe::Message tpMessage;
+  TensorpipeWriteBuffers tpBuffers;
+  std::tie(tpMessage, tpBuffers) = tensorpipeSerialize(std::move(rpcMessage));
   pipe->write(
       std::move(tpMessage),
-      // Note: keep payload and tensors of rpcMessage alive.
-      [rpcMessage{std::move(rpcMessage)},
-       reservedTensors{std::move(tpEntry.reservedTensors)},
-       copiedTensors{std::move(tpEntry.copiedTensors)},
+      [tpBuffers{
```

Comment:
Looks like the only reason we capture the `tpBuffers` here is to keep those buffers alive? If so, why not letting `tpMessage` take the ownership?

Reply:
That's true, we just capture tpBuffers to keep the memory alive. The tp::Message offers no way to take ownership of data. We used to have something like that but it turned out it's rather complicated: the original data could be managed in various forms (the ones we have here are `std::vector<char>`, `std::string` and `std::unique_ptr<...>`, but there exist so many more) and TP cannot support all of them. One way this could work is through `std::unique_ptr` as it supports a custom destructor,  so all the above ways could be "converted" to that. However that destructor must be specified as a template parameter, so in order to be generic this would have to be something like `std::unique_ptr<uint8_t, std::function<void(uint8_t*)>>`. And, at this point, I'm not sure it would really be more convenient for the agent to create a custom destructor rather than capture it in the lambda it's already creating anyways.

Reply:
I see.